platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintChecker.kt

```diff
@@ -310,17 +314,43 @@ class ConfigurationCacheFingerprintChecker(private val host: Host) {
         }
     }
 
-    private
-    fun hasFileChanged(file: File, originalHash: HashCode) =
-        !isUpToDate(file, originalHash)
-
     private
     fun hasDirectoryChanged(file: File, originalHash: HashCode?) =
         host.hashCodeOfDirectoryContent(file) != originalHash
 
     private
-    fun isUpToDate(file: File, originalHash: HashCode) =
-        host.hashCodeOf(file) == originalHash
+    fun isFileUpToDate(file: File, originalHash: HashCode) =
+        checkFileUpToDateStatus(file, originalHash) == UpToDateStatus.Unchanged
+
+    enum class UpToDateStatus {
+        Unchanged,
+        FileChanged,
+        TypeChanged,
+        Removed
+    }
+
+    private
+    fun checkFileUpToDateStatus(file: File, originalHash: HashCode): UpToDateStatus {
+        val (hashCode, fileType) = host.hashCodeAndTypeOf(file)
+        return doCheckFileUpToDateStatus(originalHash, hashCode, fileType)
```

Comment:
This "layer of inderection" seems a bit redundant (this `do` prefix is telling us), maybe you could consider to return smth like this:
```
return when {
   hash == originalHash -> UpToDateStatus.Unchanged
   fileType == FileType.RegularFile -> UpToDateStatus.FileChanged
   fileType == FileType.Directory -> UpToDateStatus. TypeChanged
   fileType == FileType.Missing -> UpToDateStatus.Removed
}
```
I don't insist, just an option